docs/api: introduce KubeVirt API design guidelines (#18612)
kubevirt/kubevirt: docs/api/design_guidelines.md

> # KubeVirt API Design Guidelines & Best Practices

Reviewer: I've audited the existing VM, VMI, and instancetype APIs against these guidelines and the gap between the stated rules and current reality is substantial. The document uses strong normative language ("must", "strictly forbidden", "must be avoided") but the existing core APIs violate many of these rules extensively:

- Zero CEL immutability rules on hardware identity fields (firmware UUID, SMBIOS serial, MAC addresses) - and no webhook enforcement on the VM resource either
- Near-zero CEL validation adoption overall (3 XValidation markers in all of core/v1, vs ~37 CEL-expressible checks in imperative webhooks)
- Widespread `uint32`/`uint64` usage across CPU topology, boot order, and config fields
- Hard `+kubebuilder:validation:Enum` markers on extensible fields (RebootPolicy, VMRolloutStrategy, InstancetypeReferencePolicy, MigrationCompression)
- Multiple phase-based state machines added after `vmi.Status.Phase` (VolumePhase, MemoryDumpPhase, ChangedBlockTrackingState)
- `float32`/`float64` fields (TokenBucketRateLimiter.QPS, LoginTime)
- `VolumeStatus.Size` using `int64` instead of `resource.Quantity`
- `QemuGuestAgent` as a field name in the public API
- References to server-side apply despite VEP 389 not yet being implemented

The document does acknowledge some existing debt explicitly (vmi.Status.Phase, custom conditions, VMI /status subresource) but many other gaps are left unaddressed - the reader has to discover for themselves that the "must" rules don't describe current reality.

I think this would benefit from a VEP first, framing these as target guidelines for new API work and providing a roadmap for retrofitting the existing APIs. As it stands, the document reads as a description of how things are when it's really a description of how things should be. A short "Current State vs. Target State" section in the introduction - or consistent "existing fields predate this rule" caveats - would prevent confusion.

I've put together a detailed audit of the findings here for reference: https://github.com/lyarwood/kubevirt/blob/api-guidelines-audit/docs/api/design_guidelines_audit.md

Author: Agreed on the framing gap. Added a "How to Read This Document" section before §1 that explicitly states these are target-state guidelines for new API work, lists the main categories of acknowledged debt, and notes that retrofitting existing violations will go through individual VEPs. A VEP for the guidelines document itself is not needed — this is a living reference, not a feature proposal, analogous to how Kubernetes maintains `api-conventions.md` outside its enhancement process.

## Table of Contents

1. [Foundations](#1-foundations)
2. [Resource Hierarchy & Lifecycle](#2-resource-hierarchy--lifecycle)
3. [Field & Type Design](#3-field--type-design)
4. [Validation](#4-validation)
5. [Hypervisor Translation Layer](#5-hypervisor-translation-layer)
6. [Status & Conditions](#6-status--conditions)
7. [References & Sources](#7-references--sources)

---

## 1. Foundations

### 1.1 Backend Agnosticism

> KubeVirt's public API must remain decoupled from any specific backend implementation — both to future-proof the project and to allow for the potential coexistence of alternative backends. This applies at every layer of the stack: hypervisor engines, storage drivers, network plugins, and future virtualization backends must not leak their names, native types, or internal semantics into the public API surface. API fields, condition types, and validation rules should use generalized, implementation-neutral nomenclature — `hypervisor`, `domain`, `runtime` — rather than engine-specific terms. The hypervisor is the most prominent example of this principle because the temptation to embed `qemu`- or `libvirt`-specific concepts is highest at the virtualization boundary; the same discipline applies equally to storage backends (e.g., avoid encoding a specific CSI driver's behavior in a field name) and network plugins.

Reviewer 2: can you break it to multiple easier to review lines?

To me,
```
The hypervisor is the most prominent example of this principle because the temptation to embed `qemu`- or `libvirt`-specific concepts is highest at the virtualization boundary; the same discipline applies equally to storage backends (e.g., avoid encoding a specific CSI driver's behavior in a field name) and network plugins
```
is to long a sentence. Would you break it to two examples - what should, and what should not happen with hypervisor and network API fields.

Author: Done — split into three bullets (hypervisor, storage backends, network plugins).

> **Justified exceptions:** A field whose explicit purpose is to identify or select a specific backend may name that backend directly. `HypervisorConfiguration.Name` (values: `kvm`, `hyperv-direct`) is the canonical example — the field exists precisely to declare which hypervisor is present. Similarly, expert or debug annotations such as `kubevirt.io/libvirt-log-filters` are deliberate escape hatches that must name the engine to be meaningful. Each such exception must be explicitly motivated in the field's API comment or the relevant [VEP](https://github.com/kubevirt/enhancements).

Reviewer: `HypervisorConfiguration.Name` is cited here as the canonical example of a justified exception to backend agnosticism, but the field in the codebase uses `+kubebuilder:validation:Enum=kvm;hyperv-direct` - a hard OpenAPI enum. Section 3.2 states that hard enum arrays "must be avoided on fields representing extensible lists of choices" because they break rolling upgrades.

This creates an unaddressed tension within the document. It might be worth acknowledging this in Section 3.2 as a known exception (with justification - e.g., adding a new hypervisor backend is a major architectural event, not an incremental extension like a new disk bus type) or noting it as existing technical debt.

Author: Good catch. After reviewing the reasoning, the enum on `HypervisorConfiguration.Name` does not hold up as a justified exception — the rolling upgrade risk applies identically regardless of whether the new value represents a minor extension or a major architectural event. Acknowledged it as technical debt in a Note in §3.2.1, along with `ExperimentalMigrationOptions.Compression` and `TLSConfiguration.MinTLSVersion` as the other clear cases. Added a grep pattern to identify additional violations.

### 1.2 Kubernetes Ecosystem Alignment (The KubeVirt Razor)

KubeVirt follows the architectural rule known as the **KubeVirt Razor**: *"If something is useful for Pods, we should not implement it only for VMs."* The intent is to prevent KubeVirt from building a parallel ecosystem that duplicates mechanisms Kubernetes already provides. KubeVirt's API should feel like a natural extension of Kubernetes, not a separate system running alongside it.

When a Kubernetes standard mechanism exists, it must be used in preference to a KubeVirt-specific equivalent:

* **Types:** Use `metav1.Condition` for conditions, `resource.Quantity` for sizes and capacities, `metav1.Duration` for durations. Introducing a KubeVirt-specific parallel requires explicit justification.

> * **Patterns:** Use owner references for garbage collection, [server-side apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) for field ownership, standard `status.observedGeneration` for controller sync signaling. The cost of deviating from a standard pattern is paid by every consumer of the API.

Reviewer: server-side apply is only just being introduced in v1.10.0 with https://github.com/kubevirt/enhancements/pull/390?

Author: Correct — moved SSA to a separate sentence, annotated as a target pattern with adoption in progress via VEP #390.

* **Conventions:** Follow Kubernetes naming, serialization, and deprecation conventions (documented in the [Kubernetes API guidelines](https://github.com/kubernetes/community/blob/main/contributors/devel/sig-architecture/api-conventions.md)). Deviations create a learning tax for users and operators already familiar with the Kubernetes ecosystem.

> KubeVirt does deviate from Kubernetes conventions in cases where virtualization semantics have no Pod equivalent — the VM→VMI orchestration hierarchy, the sticky-spec back-write for hardware identity, and the hypervisor abstraction layer are canonical examples. Each such deviation must be explicitly motivated in the relevant API documentation or [VEP](https://github.com/kubevirt/enhancements); the mere fact that something is convenient or already implemented is not sufficient justification.

Reviewer 2: `the sticky-spec back-write for hardware identity`

please add a link - I did not understand from the text that it's about a VM controller modifying the vm.spec.

Reviewer 2: can you add an good example for how this SHOULD be done.

Author: Done — added a link to §2.2 and an example of a justified deviation: the sticky-spec back-write, which deliberately crosses the spec/status boundary to persist hardware identity values the guest OS has already committed to.

Reviewer 2: thanks. But I intended my request as more general - please use shorter lines in this entire document. They are easier for humans to review and suggest modifications.

---

## 2. Resource Hierarchy & Lifecycle

### 2.1 The Desired Intent vs. Active Execution Boundary

#### 2.1.1 The VirtualMachine (VM) Contract

> The `VirtualMachine` resource represents the persistent, long-lived management object (analogous to a specialized Kubernetes `StatefulSet`). It acts as the definitive source of truth for the user's permanent desired state — including boot configuration, storage topologies, and high-level power state intent (e.g., `Running` or `Stopped`). A `VirtualMachine` resource outlives any single guest operating system boot cycle.

Reviewer 2: `Running` or `Stopped` with backticks please refer to specific in-API fields (I'm not aware of `Stopped` field o value)

Author: Done.

#### 2.1.2 The VirtualMachineInstance (VMI) Contract

The `VirtualMachineInstance` represents the transient execution span of a virtual machine runtime. Unlike standard Kubernetes `Pods` which maintain a 1:1 relationship with an active container boundary, a VMI tracks a continuous virtualization lifecycle. It must accommodate complex runtime orchestration, such as mapping to multiple underlying infrastructure primitives concurrently (e.g., managing both a source and a target Pod during Live Migration).

#### 2.1.3 The Unidirectional Mutation Loop

> As a general paradigm, users mutate the `VM` specification. The VM controller reconciles this intent by managing the creation, deletion, or highly controlled mutation of the underlying `VMI` specification.

Reviewer 2: ```suggestion
Users mutate the `VM` specification. The VM controller reconciles this intent by managing the creation, deletion, or highly controlled mutation of the underlying `VMI` specification. The effects of a user changes to a VM-controlled VMI are undefined
```
not sure if it's good, but I did not get the Paradigm. Do we have a counter-examples?

Author: Removed the "paradigm" framing. Strengthened the clarification: direct user changes to VMI spec are not undefined — they are actively rejected at the API boundary by the validating admission webhook.

---

> ### 2.2 Late Defaulting and Identity Persistence (The Sticky-Spec Rule)

Reviewer 2: ```suggestion
### 2.2 Identity Persistence Considerations
```

Author: Done.

#### 2.2.1 The Architectural Exception

> In standard cloud-native environments, restarts provide a completely blank slate. In virtualized infrastructure, however, specific hardware identifiers dictate the permanent identity of a virtual machine and must survive across hard reboots. If a guest OS unexpectedly loses its firmware UUID, SMBIOS serial number, or persistent MAC address between boot cycles, it can trigger guest OS license deactivations, alter network interface ordering, or cause catastrophic system failure.

Reviewer 2: ```suggestion
In standard cloud-native environments, restarts provide a completely blank slate. In virtualized infrastructure, however, specific hardware identifiers dictate the permanent identity of a virtual machine and must survive across hard reboots. For example, if a guest OS unexpectedly loses its firmware UUID, SMBIOS serial number, or persistent MAC address between boot cycles, it can trigger guest OS license deactivations, alter network interface ordering, or cause catastrophic system failure.
```

Author: Done.

#### 2.2.2 Identity Value Lifecycle Patterns

Some fields, such as hardware identity fields, are complex, low-level details that users should not be expected to manage. KubeVirt does not require them to be set at `VM` creation time, and instead takes care of generating and persisting these values transparently — using the patterns below — so that the VM lifecycle remains stable without user intervention.

> * **Pattern 1 — Admission Webhook Seeding (primary pattern):** For new VMs, the mutating admission webhook generates stable identity values and writes them directly into `VM.spec` at creation time. The firmware UUID is the canonical example: if not provided by the user, the webhook seeds it before the object is persisted. Anchoring in `VM.spec` ensures the value survives across the entire VM lifecycle, including reboots and re-creations of the underlying VMI.

Reviewer: This concern came up during my review of #17399 (VEP 14 Phase 2), where a commit extends the VM mutating webhook to assign firmware UUID on both Create and Update operations to close a gap in the snapshot restore path. That review prompted me to look more carefully at whether codifying Pattern 1 and Pattern 2 here is the right direction — I don't think it is.

**Kubernetes API convention: spec is user intent**

The [Kubernetes API conventions](https://github.com/kubernetes/community/blob/main/contributors/devel/sig-architecture/api-conventions.md#spec-and-status) are explicit:

> *Spec is a description of the desired state of the object as supplied by the user.*
> *Status is a description of the current state of the object, populated by the system.*

A firmware UUID or architecture value that the user did not write is not user-supplied desired state — it is a system-assigned value. The right home for that is `status`, not `spec`. The fact that we need it to survive reboots is a persistence argument, not a placement argument; status fields survive reboots just as well as spec fields, and controllers can read from status just as easily.

**GitOps breakage**

When a mutating webhook writes fields into `spec` that the user never authored, every GitOps tool (ArgoCD, Flux, etc.) will detect permanent drift between the manifests in git and the objects in the cluster. Users applying a VM definition get back a spec that has been silently augmented with a firmware UUID, architecture, machine type, and serial number they never wrote. This is a known, widely-complained-about failure mode for Kubernetes operators.

**The instancetype precedent**

KubeVirt has already walked this road and walked back from it. The instance type design went through a period where system-assigned values (`ControllerRevisionRef`, expanded fields) were written into `VM.spec`. That caused real problems — GitOps drift, unexpected diffs on update, confusing ownership semantics — and the project has been incrementally moving those values into `status` (see the `ControllerRevisionRef` in `status.instancetypeRef`, and the ongoing removal of `PatchStatus`-into-spec flows). We should not codify as a rule the pattern we have been actively undoing.

**What the guidelines should say instead**

Rather than "the webhook seeds identity values into spec (primary pattern)", I would suggest:

- System-assigned identity values (firmware UUID, serial, architecture when derived from cluster config rather than user choice) belong in `status`.
- `spec` holds only what the user explicitly authored. If the user wants to pin a firmware UUID, they write it in spec; that overrides status.
- The VM controller, during normal reconciliation, is responsible for generating and persisting these values into `status`. This is consistent with the controller-owns-status principle already described in §6.6.
- The VMI creation path reads: spec value if non-empty, otherwise status value.
- Webhooks remain thin and do not write to status; they only validate or default what the user explicitly provided.

Pattern 2 (controller back-fill into spec for migration) is the hardest to eliminate since it was specifically designed for the upgrade window problem, but even there the better model would be a one-time migration controller that writes to `status`, with the VMI creation path reading from status as the canonical fallback. The firmware UUID `FirmwareController` that was recently removed in #17399 was essentially this pattern, just targeting the wrong field (spec instead of status).

I would strongly prefer §2.2 either be removed pending a VEP that properly designs the status-based approach, or reframed to flag Pattern 1/2 as historical technical debt with a note that new identity fields should use status.

Author: Thank you for the detailed review — these are important points worth addressing carefully.

**On the Kubernetes convention (spec = user intent):**
We would characterise what §2.2 describes as *late defaulting* rather than system-generated state. The firmware UUID is a value the user *can and should* set explicitly — that is the preferred path. For users who prefer not to manage low-level hardware details, KubeVirt provides a convenient default. Where the convention must bend is when the default is complex or hypervisor-dependent and cannot be computed at object creation time (the firmware UUID upgrade path is the concrete example). This is closer to a mutating webhook default than to writing runtime-observed state into spec.

**On GitOps breakage:**
This concern applies equally to any mutating webhook that defaults spec fields — a general and well-understood pattern in Kubernetes. A GitOps user who does not want to manage the firmware UUID has explicitly chosen not to be explicit about it. GitOps tooling should respect that delegation and not reconcile fields the user chose not to own (via `ignoreDifferences` or equivalent). The discipline is: be explicit about what you manage. Fields you leave to defaults are, by definition, outside your management scope.

**On the instancetype precedent:**
A user who references an instancetype has made an explicit choice not to be explicit about CPU and memory. The same principle applies: if you delegate a decision to KubeVirt, your GitOps configuration should reflect that. The lesson from the instancetype experience is to document clearly which fields are user-managed vs system-defaulted — not that defaulting into spec is inherently wrong.

**On the proposed target architecture (status as identity store):**
This is worth discussing further and we are open to revisiting it. We do want to raise one concern: §5.2.3 and §6.3 of this document establish that `status` should reflect the current observed state of the running hypervisor and must be recomputable from it — not used as input to drive the next reconcile cycle. A firmware UUID in `status` on a stopped VM would not be observed from current hypervisor state; it would be a historical value carried forward to seed the next boot, making `status` a durable configuration store — which is the role of `spec`. We are not saying this makes the proposal impossible, but it surfaces a tension that would need to be resolved in a VEP. Happy to continue this discussion.

**On removing or reframing §2.2 as debt:**
Given the above, we believe the patterns in §2.2 are justified deliberate deviations with explicit reasoning, not accidental violations. We would prefer to keep the section as-is for now. That said, if specific points within §2.2 remain genuinely unresolved after further discussion, we would consider removing those points in favour of landing agreed content — we would rather ship a smaller set of well-agreed guidelines than block on debated ones. We hope to see additional feedback to better understand the extent of the debate.

When a new spec field is introduced, its design needs to take into account its persistence requirements. If it is identified as requiring one, the required lifecycle (VM vs VMI) must be considered upfront and an explicit persistence strategy defined.

---

### 2.3 Resource Ownership, Garbage Collection, and Finalizers

#### 2.3.1 Owner References for Parent-Child GC

Owner references express the parent-child relationship between KubeVirt resources and enable Kubernetes to cascade-delete children when their owner is deleted. The canonical chain is:

```
VirtualMachine → VirtualMachineInstance → virt-launcher Pod
```

The VM controller creates the VMI with an owner reference pointing to the VM. The VMI controller creates the virt-launcher Pod with an owner reference pointing to the VMI. When the VM is deleted, Kubernetes cascades deletion to the VMI; when the VMI is deleted, the Pod follows.

**Namespace constraint:** Owner references must be within the same namespace. Cross-namespace owner references are explicitly disallowed by Kubernetes and must never be introduced.

#### 2.3.2 Finalizers for Deletion Sequencing

> Owner references alone do not guarantee ordering. Kubernetes background deletion (the default) removes the owner immediately and cleans up dependents asynchronously — the GC controller's cache may also miss objects created concurrently with deletion. For KubeVirt, where a running hypervisor domain must be stopped and pods must be cleaned up before their parent objects disappear, this timing is not sufficient.

Reviewer 2: > the GC controller's cache may also miss objects created concurrently with deletion

I don't understand this. Can you refer to a k8s doc about this? It sounds as if k8s deletion may leak new children.

Author: Dropped that clause — the core point is that background GC offers no ordering guarantee, which is sufficient to motivate finalizers.

Finalizers are the correct tool for sequencing: they block the object's removal from etcd until the controller explicitly clears them, giving the controller a guaranteed window to complete cleanup.

KubeVirt uses two finalizers in the VM→VMI lifecycle:

* **`kubevirt.io/foregroundDeleteVirtualMachine`** on VMI — placed by the mutating admission webhook at VMI creation time; cleared by the VMI lifecycle controller only after the VMI has reached a terminal phase (Failed or Succeeded) and all associated pods have been deleted.
* **`kubevirt.io/virtualMachineControllerFinalize`** on VM — held by the VM controller; cleared after the VMI has been fully removed from etcd and the VM controller has completed its own cleanup.

> The VM controller also places `kubevirt.io/virtualMachineControllerFinalize` on the VMI itself. This is complementary to the owner reference: the owner reference triggers cascade deletion, while the additional finalizer on the child gives the VM controller an explicit gate to observe the VMI reaching a terminal state before the VMI disappears from etcd.

Reviewer 2: I don't understand why this finalizer is set on the VMI and why do you mention it here. This document should lay out design guidelines and call out well-aligned examples, and maybe acknowledge less-aligned ones. Does this doc recommend using more of this? E.g. spell out what must the VM controller do /before/ the VMI is deleted.

Author: Clarified: the finalizer gives the VM controller a guaranteed observation window before the VMI disappears from etcd. Currently that window is used to read the VMI's final status and write it onto the VM object. Explicitly cross-referenced §2.3.4 to show this follows the stated guideline.

* Add a finalizer to the parent if the parent controller must observe the child's terminal state before completing its own cleanup.
* Do not rely on Kubernetes background GC timing for correctness. If ordering matters, use finalizers.
* Never add a new finalizer to an object after its `deletionTimestamp` has been set — Kubernetes rejects this.

---

## 3. Field & Type Design

### 3.1 Kubernetes-Native Data Primitives

#### 3.1.1 Supported Primitives

The public API surface (spanning both `spec` and `status`) should utilize deterministic, Kubernetes-friendly data primitives: booleans, strings, signed integers (`int32`/`int64`), and `resource.Quantity`. Unsigned integers (`uint32`/`uint64`) exist in the API as technical debt and must not be used in new fields; where they appear, the mitigations in §3.1.2 apply.

#### 3.1.2 Mapping Rules

* **Capacities & Sizes:** Any field representing size, memory allocation, or storage limits must use `resource.Quantity`. This abstracts hypervisor byte-level or sector-level requirements away from the end user into clean, cloud-native strings (e.g., expressing memory as `2Gi` or disk size as `40Gi` instead of raw integers).

> * **Fractional or Decimal Weights:** Raw floating-point numbers are strictly forbidden within the public API `spec`, even where a backend (e.g., the hypervisor) accepts or requires a float natively. Floats introduce non-deterministic serialization behavior across different client architectures (Go, Python, JavaScript), causing values to silently change when round-tripped through different implementations. Any field requiring fractional granularity must be represented as an integer using milli-units (e.g., representing a performance weight of `1.5` as `1500m`); the conversion to the backend's native float happens at the translation layer (§5.1).

Reviewer 2: `float` should be avoided from `status`, too. You don't want to modify ETCd for every 0.001 change in a float. If you need to report a float - it's probably a metric.

If float is reported in the status, similar rounding issues can happen on the clients.

Author: Extended the prohibition to cover `status` as well. Added a Note explaining why the rule is stricter for `spec` (round-trip corruption drives controller logic) than for `status` (efficiency and display consistency). Frequent float values belong in a metrics endpoint.

* **Oversized Unsigned Values:** Hypervisor configurations that naturally use `uint64` must be mapped to safe `string` types or verified `int64` fields accompanied by rigid OpenAPI upper/lower boundary validation tags to prevent overflow exploits.

---

> ### 3.2 The Open String Set Pattern (Replacing Enums)

Reviewer 2: ```suggestion
### 3.2 The Open String Set Pattern (Avoiding Enums)
```

Author: Done.

> #### 3.2.1 The Upstream Compatibility Boundary

Reviewer 2: ```suggestion
#### 3.2.1 Kubernetes Compatibility
```

Author: Done.

#### 3.2.2 Schema Implementation

Fields that represent a selection from a set of string constants must be defined as open `string` primitives in the OpenAPI v3 schema.

#### 3.2.3 Last-Mile Validation & Graceful Rejection

Instead of hard-rejecting unrecognized values at the cluster edge, validation and error handling are deferred to the internal controller loops and node-level daemons:

* **Serialization Safety:** If an older KubeVirt component processes a resource containing a newer, unrecognized string value, it must smoothly ingest, parse, and pass the object without failing or crashing.

> * **Soft Failure Transmission:** The reconciling controller must observe the unrecognized value and gracefully pause execution *for that specific object only*. It must signal the error by updating the object's `status.conditions` to `Ready=False` with a clear programmatic code (e.g., `Reason=UnsupportedValue`) and an explanatory human-readable message.

Reviewer 2: ```suggestion
* **Soft Failure Transmission:** The reconciling controller must observe the unrecognized value and gracefully pause execution *for that specific object only*. It must signal the error by updating the object's `status.conditions` e.g. to `Ready=False` with a clear programmatic code (e.g., `Reason=UnsupportedValue`) and an explanatory human-readable message.
```

Author: Done.

---

### 3.3 Immutable Field Enforcement

#### 3.3.1 When to Apply Schema-Level Immutability

> Not every field that is persistent or sticky requires a schema-level immutability rule. CEL immutability enforcement is warranted for `VM` spec fields where a change would produce incorrect behavior in the guest OS regardless of the VM's running state — specifically hardware identity fields such as firmware UUID, SMBIOS serial number, and MAC addresses. These are distinct from:

Reviewer 2: I don't think that this section is true. For sure, we should be able to modify a MAC address of a VM, just like we can change the NIC of a physical machine.
It is less clear, but I'd say that it is fine to let a VM owner modify their firmware UUID and SMBIOS serial number. KubeVirt should protect only things that would confuse it, it should not play the bad cop handing a VM owner only reinforced machines.

Author: Good point — this raises a broader question worth discussing before settling on a rule.

A few thoughts on the specific cases:

**MAC address:** On physical hardware, changing a NIC's MAC is a discrete hardware replacement event that the guest OS observes as a device add/remove. KubeVirt models that correctly through NIC hotplug/coldplug — the guest sees a proper device lifecycle event. Direct `VM.spec` field mutation is not equivalent; it changes hardware identity without that lifecycle event, which can disrupt network configuration, bonding, and interface ordering in the guest in ways that are hard to recover from. My view is that the right path for a MAC change is hotplug (replace the NIC), not a raw spec edit, and the immutability rule enforces that.

**Firmware UUID:** This is committed to by the guest OS at first boot — UEFI variables, software licensing, TPM state, and persistent device paths all anchor to it. I do not see a valid in-place change use case. If a different firmware UUID is needed, the right operation is cloning the VM (creating a new identity), not mutating an existing one.

There is also a broader generalisation here: the immutability principle applies most clearly to **resources that represent an active runtime execution** (VMI, Pod) — those are unambiguously immutable because the hardware is live. For **management/desired-state resources** (VM), it is less clear-cut and is where your challenge applies.

Given the debate, I see two paths forward:
1. **Defer §3.3** — remove it from this PR, resolve the question in a follow-up once there is community agreement.
2. **Resolve it now** — agree on a direction in this thread and update the section accordingly.

I would lean toward option 2 if we can reach agreement here, but I am open to option 1 if this needs broader input. What is your preference?

Reviewer 2: It is perfectly fine to replace a NIC of a turned-of VM. Hence, MAC should be mutable on the API level. (You must never ever tie immutability to the Running state of the VM, TOCTOU!)
You can track the business logic (if the VM is running, and MAC was changed, do a hotplug) in the controller logic, not in the API validation. 

It is harder to prove for other fields, but still my opinion: CEL validations - must be limited to laws-of-nature (no negative CPUs) and IEEE standards (no Hebrew in UUIDs). They should not be used to enforce business logic that may change in the future.

I hope we can find the language to resolve it now.

Author: > CEL validations - must be limited to laws-of-nature (no negative CPUs) and IEEE standards (no Hebrew in UUIDs). They should not be used to enforce business logic that may change in the future.

This is what I tried to express and link the ability to change some fields to "laws-of-nature", when comparing it to a BM machine.
In the real-world, we cannot change the MAC address of a HW nic or the firmware ID. It happens only if we change the HW itself, i.e. unplug/plug it.
In my mind, I would prefer not to allow changing the mac address once it is set. If a user wants to change it, it should remove it and add another, mimicking the reality.

IMO, opening some cases to mutation triggers side effects which as operators and support, raises the complexity.

I will go ahead and defer this point to a follow up (remove it from this PR and opening a different PR).

Author: Section removed. The content is preserved in [kubevirt/kubevirt#18811](https://github.com/kubevirt/kubevirt/issues/18811) for future resolution.

---

## 4. Validation

### 4.1 Structural Invariants (Simple Field Constraints)

#### 4.1.1 The Extensibility Rule

OpenAPI schema-level constraints — such as `Minimum`, `Maximum`, `MaxLength`, `MaxItems`, and `Pattern` — must **only** be enforced if they represent an absolute physical, mathematical, or architectural invariant that is fundamentally incapable of changing in the future. Arbitrary boundaries established for business logic or aesthetic neatness artificially restrict system growth and require disruptive, coordinated CRD schema rollouts to alter.

#### 4.1.2 Legitimate Structural Invariants

Schema constraints are strictly reserved for unyielding specifications, including:

* **Mathematical Primitives:** `// +kubebuilder:validation:Minimum=1` for virtual CPU counts (a runtime environment cannot execute on zero or negative processing cores).

> * **Network & Protocol Specs:** `// +kubebuilder:validation:Maximum=65535` for standard TCP/UDP port allocations (bounded firmly by the 16-bit internet protocol standard).

Author: Done.

* **Fixed Cryptographic/IEEE Formats:** `// +kubebuilder:validation:Pattern=...` for MAC addresses or standard UUID structures where the string layout is governed by immutable external RFC standards.

> #### 4.1.3 The Safety Ceiling Pattern (Preventing etcd Exploits)

Reviewer 2: must we really protect etcd? Doesn't it have its own limits on CR sizes?

Author: You are right — etcd has its own ~1.5MB hard limit per object, so "protecting etcd" was an inaccurate framing. The real motivation is per-field: a single string field has no automatic limit and could consume most of the object budget without a clear error. Updated the section to reflect this and dropped the misleading heading parenthetical.

Reviewer 2: I don't think that it is our business to bother about that, still. Does any k8s CR have a similar provision for this purpose?

Author: You are right — Kubernetes CRDs do not add per-field `MaxLength` for etcd-safety purposes, and the framing was wrong. Dropping §4.1.3.

That said, there is a narrower case worth tracking: fields explicitly designed to carry large user-supplied content (e.g., cloud-init `userData`, inline scripts) are qualitatively different from name/identifier fields — a generous ceiling there would document the intended operating range and surface a clear error before an object silently exhausts the etcd budget. We have opened https://github.com/kubevirt/kubevirt/issues/18811 to track this and other follow-up items from the guidelines review.

* *Best Practice:* Omit tight validation at the schema edge, or set an expansive safety ceiling (e.g., `MaxLength=2048`) to protect etcd buffer limits. Any specific business-logic constraints must be evaluated flexibly within the controller runtime or soft validation blocks where they can be dynamically updated without breaking schema compatibility.

---

### 4.2 Server-Side Validation via Common Expression Language (CEL)

#### 4.2.1 Webhook Minimization

Cross-field validation — where the validity of one configuration parameter depends directly on the state of another — must prioritize high-performance **CEL (Common Expression Language)** rules via the `+kubebuilder:validation:XValidation` [marker](https://book.kubebuilder.io/reference/markers/crd-validation), which instructs Kubebuilder to generate the corresponding CEL [validation rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules) in the CRD schema. Writing or extending imperative validating admission webhooks should be treated as a secondary fallback reserved exclusively for highly dynamic, external database lookups.

#### 4.2.2 Application Examples

> CEL rules allow immediate server-side rejection at the API boundary before etcd persistence, saving critical controller compute cycles:

Reviewer 2: ```suggestion
CEL rules allow immediate server-side rejection at the API boundary before etcd persistence, saving controller compute cycles:
```
please avoid hyperbolic language.

Author: Done

* **Exactly One:** Ensuring that a virtual disk drive slot does not simultaneously define conflicting storage backends (illustrative — the real `Disk` struct has many more backend types; a production rule must account for all of them):
  `// +kubebuilder:validation:XValidation:rule="has(self.containerDisk) != has(self.persistentVolumeClaim)",message="A disk slot must define exactly one backend provider (ContainerDisk or PVC)"`

> * **Dependency Enforcement:** Verifying that a hypervisor feature flag is only supplied if its pre-requisite device topology is explicitly allocated.

Reviewer 2: I don't understand this example.

Author: The `!=` on two booleans is a CEL XOR — `true` when exactly one field is present, `false` when both or neither are. Added an explanation sentence inline before the example.

Reviewer 2: I still don't follow. 
```
* **Dependency Enforcement:** Verifying that a hypervisor feature flag is only supplied if its pre-requisite device topology is explicitly allocated.
```

maybe you can give a yaml example of good and rejected values

Author: I am easing the wording and give a simple example.

  This pattern carries a hard design constraint: **the zero or absent value of the field must be a safe no-op for every consumer.** If any consumer acts on the zero value as though it were meaningful data, it will break before the gate is ever enabled — silently, with no error. For example, a consumer checking condition freshness with `condition.ObservedGeneration < vm.Generation` would evaluate `0 < 5 = true` and incorrectly treat every condition as stale on every VM, regardless of actual state. The safe consumer pattern is to treat zero/absent as "information not yet available" and skip the check entirely:

  ```go
  if condition.ObservedGeneration > 0 && condition.ObservedGeneration < vm.Generation {
      // treat condition as stale
  }
  ```

>   To illustrate why zero-safety matters beyond gate toggles: if a hypothetical downgrade were to land a cluster on a release that predates the gate (KubeVirt does not support rollbacks, but the scenario is instructive), controllers would stop populating the field and consumers would again observe zero/absent. A consumer written without the `> 0` guard would misfire on every VM. Write the guard from day one, before the gate is ever enabled.

Reviewer 2: I don't find this hypothetical example helpful. Would it make sense to speak about enabling a feature gate and then disabling it?

Author: Good suggestion. Replaced the hypothetical downgrade with a feature gate disable scenario (e.g. disabling a gate while investigating a regression), which is a real situation that actually occurs.

The translation of safe, standardized Kubernetes API primitives into raw hypervisor-specific types (e.g., converting a `resource.Quantity` into a raw bit-mask, byte count, or string configuration block, or a milli-unit integer into a floating-point number) must be deferred to the absolute edge of execution. This translation must occur at the node level (e.g., within the localized runtime agent container), immediately before the hypervisor's native execution engine API or configuration writer is invoked.

### 5.2 The Symmetric Read-Path Translation (Hypervisor → Status)

#### 5.2.1 Status Symmetry

The abstraction layer must work symmetrically in reverse. Raw hardware metrics, execution states, and telemetry data read directly from the hypervisor engine will naturally contain floats, raw bytes, or non-standard types.

#### 5.2.2 Dynamic Ingestion

These values must be translated *back* into safe, standardized Kubernetes-native primitives at the node level immediately upon collection by the node daemon. The data published to the `VMI.status` block must strictly mirror the data types, granularities, and structures established in the `spec`.

#### 5.2.3 Hypervisor as Source of Truth

> The node-level daemon must be capable of querying the active hypervisor provider on the host to rebuild the VMI `status` completely from scratch. If the control plane splits or etcd experiences data loss, the absolute state of truth resides within the running hypervisor instance. The control plane must be capable of perfectly reconstructing the object's `status` purely from that observed physical reality without relying on cached historical flags.

Reviewer 2: ```suggestion
The node-level daemon (virt-handler) must be capable of querying the active hypervisor provider on the host to rebuild the VMI `status` completely from scratch. The absolute state of truth resides within the running hypervisor instance, not in caches, remote controllers or etcd.
```
I don't think that etcd data loss is a true reason. The important point is that anything but the local hyperversor is out of date and raceful.

Author: Agreed — the failure scenario framing was a distraction. The real point is that the running hypervisor is always the source of truth; everything else is potentially stale or racy. Updated accordingly.

> ### 5.3 Reconciliation Integrity

Reviewer 2: this section is bizarre. I don't follow it at all.

Author: Rewrote the section to make the concrete failure mode explicit: if the translation layer converts to hypervisor-native floats during comparison rather than only at write time, floating-point precision loss creates a spurious diff, triggering an unnecessary reconcile, which writes back, triggering another — an infinite cascade driven by float noise rather than a real config change. Also clarified the level-based model upfront so the rule makes sense in context.

Reviewer 2: I still don't really get it. It may be the dramatic title "Reconciliation Integrity" that blows me off track.

Author: Controllers are reconciling in a loop based on the spec and current status and other runtime information.

If the reconciliation is going to frequently require changing the status or perform other runtime actions, it is causes a storm of changes, and distabilizes the system in general (e.g. overloading the API server, triggering other controllers actions).

Let me try and see if I can rephrase it or just move this to other better sections.

Author: Removed §5.3 entirely. The level-based reconciliation model is already covered by §6.3; the float prohibition and its reconcile-loop consequence are already covered by §3.1.2; and the write-time-only constraint is already implied by §5.1 (translate at the absolute edge, immediately before the hypervisor write). No unique content remained.

> #### 5.3.1 Level-Triggered Stability

Reviewer 2: I'm not familiar with this term. Please explain or add a URL.

Author: Added an inline explanation: "level-based" means acting on the current value of state (the level), as opposed to edge-triggered systems that react to state transitions. Linked to the definition in the Kubernetes API conventions doc which uses the term directly: "The system is level-based rather than edge-triggered."

To guarantee level-triggered stability and eliminate infinite update cascades, KubeVirt reconciliation loops must only ever compare Kubernetes-native types against Kubernetes-native types.

#### 5.3.2 Eliminating Drift

By ensuring that both the desired state (`spec`) and the translated observed state (`status`) use identical, high-precision Kubernetes primitives (such as `resource.Quantity` comparisons), the project prevents non-deterministic float drift or serialization discrepancies from falsely triggering the controller plane into accidental modification cycles.

---

## 6. Status & Conditions

> ### 6.1 Condition Exposure & The Rejection of Linear State-Machines

Reviewer 2: ```suggestion
### 6.1 Use Condition, not Linear State-Machines
```
I'd prefer down-to-earth title, with less pathos and more clarity.

Author: Done

Reviewer 2: not just here, please. Please for every section, please choose a clear and short title, that would give me the main takeout or the main subject.

Author: Done — applied a systematic pass over all section titles: removed "The" prefixes, replaced verbose noun phrases with the main takeout, and shortened throughout. Examples: "The Desired Intent vs. Active Execution Boundary" → "VM vs VMI Roles", "Server-Side Validation via Common Expression Language (CEL)" → "CEL Validation", "Future-Proofing Status Details via Abstraction" → "Abstract Status Fields", etc. Also renamed "sticky-spec back-write" to "sticky-spec late-defaulting" in the body text, consistent with the framing in the §2.2 thread.

#### 6.1.1 High-Level Orthogonal Facets

Condition types inside the `status.conditions` array must exclusively represent high-level, long-running, and orthogonal facets of a resource's readiness, capability, or health (e.g., `Ready`, `LiveMigratable`, `Paused`).

#### 6.1.2 State-Machine Avoidance (No Phase Strings)

> Do not introduce `Phase` string fields or linear state-machine tracking in new API resources. A single `Phase` string forces a multi-faceted distributed system into an imperative, edge-triggered state timeline that is prone to deadlocks and is difficult to extend without breaking consumers.

Reviewer 2: Add a URL to k8s saying this.

Author: Linked `Phase` to the "Typical status properties" section of the Kubernetes API conventions doc, which explains the preference for conditions over phase-based models.

* **Abstract Categorization:** `Reason` codes should be structured abstractly enough to group comparable operational states under broad, predictable categories (e.g., `ResourceAllocationError`, `NetworkBindingFailed`). This allows consuming controllers to make safe, generalized automation decisions.
* **Specialized Exception:** Highly focused, narrow `Reason` codes are acceptable *only* if a consuming controller needs to execute a distinctly different, automated remediation pathway for that specific sub-fault.
* **No Multi-Reason Packing:** A condition must never attempt to combine multiple distinct failures into a single programmatic reason string (e.g., avoid `StorageAndNetworkNotReady`).

#### 6.2.2 The Human Consumer Contract (`Message`)

The `Message` field is a dynamic, human-readable string intended for end-user triage, CLI output (`kubectl describe`), and log analysis. All granular multi-faceted error strings, raw hardware stack traces, nested error codes, or contextual virtualization parameters must be offloaded entirely to the `Message` field.

---

> ### 6.3 Level-Triggered Status Reflection (No History Caching)

Reviewer 2: ```suggestion
### 6.3 Level-Triggered Status
```
simpler, clearer titles would make this more readable.

Author: Done

#### 6.5.2 Consumer Generation-Awareness Contract

Any sub-component, webhook, or external operator consuming a KubeVirt resource status is not strictly forced to block execution until the status stabilizes (`metadata.generation == status.observedGeneration`). Instead, consumers must execute a conscious, contextual design decision:

* **Synchronized/High-Stakes Logic:** If the consuming logic executes an irreversible or destructive action (e.g., triggering a hard resource eviction or declaring a terminal failure state), it should check for a generation mismatch. If `metadata.generation > status.observedGeneration`, it should recognize the status as stale, defer execution, and wait for the reconciling controller to catch up.
* **Speculative/Low-Stakes Logic:** If the consumer is executing non-destructive monitoring, speculative routing, or path optimization, it may choose to act immediately on the current data, fully aware of the active reconciliation gap.

---

> ### 6.6 The Single-Owner Status Boundary (Leaf Field Isolation)

Reviewer 2: ```suggestion
### 6.6 `Status` Ownership
```
What does "Boundary" mean? Is the tree Leaf analogy commonplace?

Author: Simplified to `### 6.6 Status Ownership`.

I wanted to emphasize that a subtree leaf (group of fields) could be own by one controller while another by a different one, it does not need to be per field. As long as they are different, it is fine. In practice, this is enforced by convention or by a k8s mechanism (server side apply).

Reviewer 2: In graph theory, a "leaf" is a vertex with only one connection not a group of fields. https://github.com/kubevirt/kubevirt/pull/18612/changes#diff-ed1897aa85abceb88d0ea13ebbf17c0edbe129222cff2721bbadec319c05aedfR394 is still confusing.

Author: Renamed to "Field-Group Isolation" throughout. Also softened the ownership boundary wording to reflect that this is a convention today, enforced by SSA where available.

#### 6.6.1 The Ideal: Single-Owner Per Field

> To eliminate etcd write conflicts, optimistic concurrency failures, and infinite controller update cascades, a single status field path should be owned by exactly **one** dedicated controller binary. For new resources, this must be enforced structurally by declaring a [`/status` subresource](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#status-subresource) in the CRD. With a status subresource in place, the API server silently discards status fields included in a spec (main-resource) write — status can only be updated via the `/status` subresource endpoint — preventing spec writes from accidentally clobbering observed state. Designing single-ownership in from the start is always preferable to relying on convention.

Reviewer 2: ```suggestion
To eliminate etcd write conflicts, optimistic concurrency failures, and infinite controller update cascades, the status should be owned by exactly **one** controller binary. For new resources, this must be enforced structurally by declaring a [`/status` subresource](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#status-subresource) in the CRD. With a status subresource in place, the API server silently discards status fields included in a spec (main-resource) write — status can only be updated via the `/status` subresource endpoint — preventing spec writes from accidentally clobbering observed state. Designing single-ownership in from the start is always preferable to relying on convention.
```

The ideal is that the entire status is owned by a specific controller. Discussing specific fields is only a mitigation, to be left to the next sections. P.S the binary can handle multiple CRs, it does not have to be dedicated to a specific one.

Author: Thank you, I tried to phrase and organize it better.

#### 6.6.2 The Architectural Reality on VMI

> `VirtualMachineInstance` sits at the boundary between two distinct control planes: the VMI controller (cluster-level orchestration) and `virt-handler` (node-level domain state). Both must write to `vmi.Status` as part of their core responsibilities, and VMI does not have a `/status` subresource. This is a known architectural tension that cannot be resolved without significant rework. It is not a pattern to emulate in new resources; it is a constraint inherited from the foundational design of the platform.

Reviewer 2: ```suggestion
`VirtualMachineInstance` sits at the boundary between two distinct control planes: the VMI controller (cluster-level orchestration) and `virt-handler` (node-level domain state). Both currently write to `vmi.Status` as part of their core responsibilities, and VMI does not have a `/status` subresource. This is a known architectural tension that cannot be resolved without significant rework. It is not a pattern to emulate in new resources; it is a constraint inherited from the foundational design of the platform.
```
I don't believe it's a "must". It is a sad reality that should be fixed. Every access of virt-handler to the kube-api can be replaced by vmi controller collecting information from virt-handler and writing it to vmi.status.

Author: Agreed — updated to "Both currently write to `vmi.Status`". The existing dual-write is a known architectural constraint, not a design prescription.

Reviewer 2: I'd say, "architectural mistake" rather than "architectural constraint"...

Author: Done — used "architectural debt" rather than "mistake" to keep the framing neutral.

* **Explicit ownership documentation:** The field's API comment must state which component owns which portion of the value. This is a hard requirement — undocumented shared ownership is indistinguishable from a bug.
* **Read-before-write with preservation:** Each component reads the current field value, calculates only its own portion, and merges the result back without touching the portion owned by the other component. Neither component may assume the field is empty on write.
* **Single batched patch:** All status changes a component needs to make in a given reconcile must be collected and issued as a single patch operation. Issuing multiple sequential patches increases the window for the other component to interleave a write between them, compounding the race. Full object `.Update()` calls are prohibited — a full update from a stale in-memory copy will silently overwrite whatever the other component last wrote.

The careful merge pattern is inherently racy and error-prone. It must be treated as a last resort, applied only when partitioning the field is not possible, and subjected to heightened scrutiny in design review and code review.

---

### 6.7 Condition Design Guidelines

> #### 6.7.1 The Decision Sequence for Adding a New Condition

Reviewer 2: ```suggestion
#### 6.7.1 Adding a New Condition
```

Author: Done

> Before introducing a new condition type, apply this sequence in order:

Reviewer 2: ```suggestion
Before introducing a new condition type, apply these considerations in order:
```
"sequence" sounds mechanical.

Author: Done

#### 4.2.2 Application Examples

CEL rules allow immediate server-side rejection at the API boundary before etcd persistence:

* **Exactly One:** Ensuring that a virtual disk drive slot does not simultaneously define conflicting storage backends (illustrative — the real `Disk` struct has many more backend types; a production rule must account for all of them). In CEL, `has(a) != has(b)` is a boolean XOR — it evaluates to `true` when exactly one of the two fields is present, and `false` when both or neither are present:
  `// +kubebuilder:validation:XValidation:rule="has(self.containerDisk) != has(self.persistentVolumeClaim)",message="A disk slot must define exactly one backend provider (ContainerDisk or PVC)"`
* **Dependency Enforcement:** Verifying that a hypervisor feature flag is only supplied if its pre-requisite device topology is explicitly allocated.

> #### 4.2.3 The Rolling Upgrade Safety Boundary

Reviewer 2: Please pardon my language, but the titles still sound "pompous". Would I be wrong to guess that they were proposed by AI? Why isn't it just
```suggestion
#### 4.2.3 Upgrade Safety
```
?

Author: Done — applied your suggestion: `#### 4.2.3 Upgrade Safety`. Also used this as the starting point for a systematic title-simplification pass across all sections (see the other thread).

The translation of safe, standardized Kubernetes API primitives into raw hypervisor-specific types (e.g., converting a `resource.Quantity` into a raw bit-mask, byte count, or string configuration block, or a milli-unit integer into a floating-point number) must be deferred to the absolute edge of execution. This translation must occur at the node level (e.g., within the localized runtime agent container), immediately before the hypervisor's native execution engine API or configuration writer is invoked.

### 5.2 The Symmetric Read-Path Translation (Hypervisor → Status)

#### 5.2.1 Status Symmetry

The abstraction layer must work symmetrically in reverse. Raw hardware metrics, execution states, and telemetry data read directly from the hypervisor engine will naturally contain floats, raw bytes, or non-standard types.

#### 5.2.2 Dynamic Ingestion

These values must be translated *back* into safe, standardized Kubernetes-native primitives at the node level immediately upon collection by the node daemon. The data published to the `VMI.status` block must strictly mirror the data types, granularities, and structures established in the `spec`.

#### 5.2.3 Hypervisor as Source of Truth

> The node-level daemon must be capable of querying the active hypervisor provider on the host to rebuild the VMI `status` completely from scratch. The running hypervisor instance is always the absolute source of truth — etcd, remote controllers, and cached state are all potentially stale or racy relative to the current hardware reality. The control plane must be capable of perfectly reconstructing the object's `status` purely from that observed physical reality without relying on cached historical flags.

Reviewer 2: ```suggestion
Virt-handler must be capable of querying the active hypervisor provider on the host to rebuild the VMI `status` completely from scratch. The running hypervisor instance is always the absolute source of truth — etcd, remote controllers, and cached state are all potentially stale or racy relative to the current hardware reality. The control plane must be capable of perfectly reconstructing the object's `status` purely from that observed physical reality without relying on cached historical flags.
```
Less dramatic, more clear.

Author: Done.

### 5.3 Reconciliation Integrity

#### 5.3.1 Level-Based Stability

KubeVirt controllers use a [level-based](https://github.com/kubernetes/community/blob/03ffbd014976893b33f750b469fd6843c4070275/contributors/devel/sig-architecture/api-conventions.md#L486) reconciliation model: at each reconcile cycle the controller compares the full current desired state (`spec`) against the full current observed state (`status`), rather than reacting to individual change events. "Level-based" means acting on the current value of state (the level), as opposed to edge-triggered systems that react to state transitions. This makes controllers self-healing — a missed event, a crashed controller, or a partial status write all resolve naturally on the next cycle.

> To preserve this property, KubeVirt reconciliation loops must only compare Kubernetes-native types against Kubernetes-native types.

Reviewer 2: does this sentence belong to the next subsection?

Author: Section removed.

#### 5.3.2 Eliminating Drift

When both sides of a comparison use the same high-precision Kubernetes primitive — for example, a `resource.Quantity` in `spec` compared to a `resource.Quantity` in `status` — the result is stable and deterministic. A `2Gi` quantity compared to a `2Gi` quantity produces no diff.

> If the translation layer converts to hypervisor-native types (e.g. floats or raw byte counts) during comparison rather than only at write time, floating-point precision loss can introduce a spurious diff where none actually exists. This triggers an unnecessary reconcile, which translates and writes back, triggering another — an infinite update cascade driven by float noise, not a real configuration change.

Reviewer 2: I get `float`, but not raw bytes - they are still precise `int`. Can you provide an example beyond float?

Author: I think sticking to "float" is better.
Raw bytes fits counters & telemetries, which we have already classified as fitting serviceability and no to be handled through reconciliation (on the status).

  > (typically a single Go binary), so the round-trip corruption risk is lower;
  > the remaining concerns are efficiency (spurious etcd writes) and client-side
  > display inconsistency.

- **Oversized Unsigned Values:** Hypervisor configurations that naturally use
  `uint64` must be mapped to safe `string` types or verified `int64` fields
  accompanied by rigid OpenAPI upper/lower boundary validation tags to prevent
  overflow exploits.

---

### 3.2 Open String Set (Avoiding Enums)

#### 3.2.1 Kubernetes Compatibility

In accordance with [Kubernetes
`api_changes.md`](https://github.com/kubernetes/community/blob/main/contributors/devel/sig-architecture/api_changes.md),
the use of hard OpenAPI `enum` arrays for schema validation must be avoided on
fields representing extensible lists of choices (e.g., disk bus types, network
binding modes, hypervisor features, or power states). Hard enums break rolling
upgrades: if a newer version of KubeVirt introduces a new string token, any
older client or older control plane component running the historical schema will
completely reject the payload at the API boundary, causing serialization and

> upgrade failures.

Reviewer 3: I dont think this is the case, clients the schema validation is done on server side (k8s API server). 
So that once CRD is updated supporting new enum value, old client could send the request with the newly unknown value, server will verify the request against the schema, and approve it.
@EdDev is it not? what do I miss?

Author: A (JSON) schema is validated by the ones who use it, be it client or server.
You can find references about the difference between `validation` and `xvalidation` (CEL) in this document. Only CEL provides server side validation.

> **Note:** Several existing fields carry `+kubebuilder:validation:Enum` markers
> in violation of this rule. The clearest cases are
> `HypervisorConfiguration.Name` (`kvm;hyperv-direct`),
> `ExperimentalMigrationOptions.Compression` (`none;zstd`), and
> `TLSConfiguration.MinTLSVersion` (`VersionTLS10`–`VersionTLS13`) — all
> enumerate technology lists where new values are a natural extension path.
> Additional violations exist; search for `+kubebuilder:validation:Enum` in
> `staging/src/kubevirt.io/api/` to identify them. None of these must be used as
> a model for new fields.

#### 3.2.2 Schema Implementation

Fields that represent a selection from a set of string constants must be defined
as open `string` primitives in the OpenAPI v3 schema.

#### 3.2.3 Graceful Rejection

Instead of hard-rejecting unrecognized values at the cluster edge, validation
and error handling are deferred to the internal controller loops and node-level

> daemons:

Reviewer 4: This feels a bit too broad. IIUC the real reason to defer is **node/version skew**: during upgrade an older virt-handler can receive a value a newer virt-api accepted, and correctness then depends on the node's version - that's what the edge can't cover and what mandates graceful ingest + `status.conditions`. Worth stating explicitly.

But I'd avoid framing it as edge-vs-controller: self-contained, version-stable invariants should still be validated at the schema/webhook for fail-fast UX. How about "defer to controllers when validity depends on the runtime/node version - otherwise validate at the edge," rather than a blanket "defer all validation"?

- **Serialization Safety:** If an older KubeVirt component processes a resource
  containing a newer, unrecognized string value, it must smoothly ingest, parse,
  and pass the object without failing or crashing.
- **Soft Failure Transmission:** The reconciling controller must observe the
  unrecognized value and gracefully pause execution *for that specific object
  only*. It must signal the error by updating the object's `status.conditions`,
  e.g. `Ready=False` with a clear programmatic code (e.g.,
  `Reason=UnsupportedValue`) and an explanatory human-readable message.

---

## 4. Validation

### 4.1 Structural Invariants

#### 4.1.1 Only Absolute Invariants

OpenAPI schema-level constraints — such as `Minimum`, `Maximum`, `MaxLength`,
`MaxItems`, and `Pattern` — must **only** be enforced if they represent an
absolute physical, mathematical, or architectural invariant that is

> fundamentally incapable of changing in the future. Arbitrary boundaries

Reviewer 4: Since we now advocate for schema-level CEL validations (and not VAP objects), I wonder, why are CEL expressions (via `+kubebuilder:validation:XValidation`) can express policy restrictions that might change over time but Enums (via `+kubebuilder:validation:Enum`) must not?

Both XValidation and Enum/Minimum/Pattern can encode a constraint that's later relaxed - and relaxing is upgrade-safe (not downgrade-safe, although we don't support it IIUC). So why must schema constraints be absolute invariants?

  `Disk` struct has many more backend types; a production rule must account for
  all of them). In CEL, `has(a) != has(b)` is a boolean XOR — it evaluates to
  `true` when exactly one of the two fields is present, and `false` when both or
  neither are present:
  `// +kubebuilder:validation:XValidation:rule="has(self.containerDisk) != has(self.persistentVolumeClaim)",message="A disk slot must define exactly one backend provider (ContainerDisk or PVC)"`

- **Dependency Enforcement:** Verifying that a feature flag is only set when its
  prerequisite is also configured. For example, `isolateEmulatorThread` requires
  `dedicatedCPUPlacement`:

  ```yaml
  # Valid
  spec:
    domain:
      cpu:
        dedicatedCPUPlacement: true
        isolateEmulatorThread: true

  # Rejected: isolation requires dedicated CPU placement
  spec:
    domain:
      cpu:
        dedicatedCPUPlacement: false
        isolateEmulatorThread: true
  ```

  ```go
  // +kubebuilder:validation:XValidation:rule="!has(self.isolateEmulatorThread) || !self.isolateEmulatorThread || self.dedicatedCPUPlacement",message="isolateEmulatorThread requires dedicatedCPUPlacement"
  ```

> #### 4.2.3 Upgrade Safety

Reviewer 4: Should we advocate for the best practice being to be very restrictive at first, then relax if needed in the future, since it's easier to remove validations than to add them in terms of being upgrade-safe?

in the cluster *before* the corresponding controller binaries are fully rolled
out, developers must observe strict safety rules when introducing rules:

- **Safe CEL Usage (New Fields & Pure Invariants):** It is safe to apply CEL
  constraints to newly introduced API fields or to validate absolute, unchanging
  architectural invariants. Older active clients will not be interacting with
  these new properties and are thus insulated from validation triggers.

- **Unsafe CEL Usage (Retroactive Restrictions):** Developers must not introduce
  a CEL rule that retroactively tightens validation criteria on an *existing,
  older field* if an active legacy client or third-party operator could still be
  writing payloads that would violate the new rule — doing so breaks cluster
  state continuity and triggers abrupt, unexpected rejections during the rolling
  upgrade window. **Exception:** Kubernetes 1.28+ supports a CEL evaluation mode
  where a constraint is only checked when the field's value actually changes;
  existing objects with unchanged values continue to pass regardless of whether
  they satisfy the new rule. When the minimum supported cluster version allows

>   it, this mode is the safer way to tighten validation on an existing field.

Reviewer 4: Should we advocate for the best practice being to be very restrictive at first, then relax if needed in the future, since it's easier to remove validations than to add them in terms of being upgrade-safe?

Maybe it'd be clearer to write this directly?

>   it, this mode is the safer way to tighten validation on an existing field.

Reviewer 4: IIUC this is actually the default behavior from 1.29+. As written in the docs [1]:

> tarting with Alpha 2 in Kubernetes 1.29, x-kubernetes-validations are ratcheted only if they do not refer to oldSelf.
>
> Transition Rules are never ratcheted: only errors raised by rules that do not use oldSelf will be automatically ratcheted if their values are unchanged.

So I think if you just write a standard CEL rule and an old object in the database breaks it, k8s will automatically skip this validation during an update, as long as the user didn't modify that specific field.

It could (?) still be problematic though for users who change the value then change it back to the now-not-allowed value.. but I'm not sure how we should treat that

[1] https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules

  as absent.

  This pattern carries a hard design constraint: **the zero or absent value of
  the field must be a safe no-op for every consumer.** If any consumer acts on
  the zero value as though it were meaningful data, it will break before the
  gate is ever enabled — silently, with no error. For example, a consumer
  checking condition freshness with
  `condition.ObservedGeneration < vm.Generation` would evaluate `0 < 5 = true`
  and incorrectly treat every condition as stale on every VM, regardless of
  actual state. The safe consumer pattern is to treat zero/absent as
  "information not yet available" and skip the check entirely:

  ```go
  if condition.ObservedGeneration > 0 && condition.ObservedGeneration < vm.Generation {
      // treat condition as stale
  }
  ```

  To illustrate why zero-safety matters beyond initial enablement: if a feature
  gate is later disabled — for example, while investigating a regression or
  waiting for a hotfix — controllers will stop populating the field and
  consumers will again observe zero/absent. A consumer written without the `> 0`
  guard would misfire on every VM. Write the guard from day one, before the gate

>   is ever enabled.

Reviewer 4: What about features that are GAed for a while? Would this still be relevant?

# KubeVirt API Design Guidelines & Best Practices

## Table of Contents

1. [Foundations](#1-foundations)
2. [Resource Hierarchy & Lifecycle](#2-resource-hierarchy--lifecycle)
3. [Field & Type Design](#3-field--type-design)
4. [Validation](#4-validation)
5. [Hypervisor Translation Layer](#5-hypervisor-translation-layer)
6. [Status & Conditions](#6-status--conditions)
7. [References & Sources](#7-references--sources)

---

## How to Read This Document

These guidelines describe the **target state** for KubeVirt's API — the rules
that new API work must follow. The existing VM and VMI APIs predate many of
these rules and contain acknowledged technical debt: `uint32`/`uint64` fields,
historical `+kubebuilder:validation:Enum` markers on extensible fields,
near-zero CEL adoption, and custom condition structs without per-condition

> `observedGeneration`. Where the document calls out known gaps explicitly, those

Reviewer 2: ```suggestion
near-zero CEL adoption, custom condition structs without per-condition
`observedGeneration`, etc. Where the document calls out known gaps explicitly, those
```
I'm sure there's more tech debt. there are only examples.

Author: Done.

sections note it; the absence of such a note does not imply the existing API is
compliant.

New fields, new resources, and new API surface must follow these guidelines.
Retrofitting existing violations will be addressed through individual VEPs as
the API evolves.

---

## 1. Foundations

### 1.1 Backend Agnosticism

KubeVirt's public API must remain decoupled from any specific backend
implementation — both to future-proof the project and to allow for the potential
coexistence of alternative backends. This applies at every layer of the stack:
hypervisor engines, storage drivers, network plugins, and future virtualization
backends must not leak their names, native types, or internal semantics into the
public API surface. API fields, condition types, and validation rules should use
generalized, implementation-neutral nomenclature — `hypervisor`, `domain`,
`runtime` — rather than engine-specific terms. This applies at every abstraction
layer:

- **Hypervisor:** Avoid embedding `qemu`- or `libvirt`-specific concepts. Use
  `hypervisor`, `domain`, or `runtime` instead of surfacing engine internals.

>   The temptation is highest at the virtualization boundary, which is why the

Reviewer 2: ```suggestion
  The temptation to break this rule is highest at the virtualization boundary, which is why the
```

Author: Done.

  sizes and capacities, `metav1.Duration` for durations. Introducing a
  KubeVirt-specific parallel requires explicit justification.
- **Patterns:** Use owner references for garbage collection and standard
  `status.observedGeneration` for controller sync signaling. [Server-side
  apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) for
  field ownership is the Kubernetes-standard target pattern; KubeVirt adoption
  is in progress via [VEP
  #390](https://github.com/kubevirt/enhancements/pull/390). The cost of
  deviating from a standard pattern is paid by every consumer of the API.
- **Conventions:** Follow Kubernetes naming, serialization, and deprecation
  conventions (documented in the [Kubernetes API
  guidelines](https://github.com/kubernetes/community/blob/main/contributors/devel/sig-architecture/api-conventions.md)).
  Deviations create a learning tax for users and operators already familiar with
  the Kubernetes ecosystem.

KubeVirt does deviate from Kubernetes conventions in cases where virtualization
semantics have no Pod equivalent — the VM→VMI orchestration hierarchy, the
[sticky-spec late-defaulting](#22-identity-persistence-considerations) for

> hardware identity, and the hypervisor abstraction layer are canonical examples.

Reviewer 2: ```suggestion
hardware identity, and the hypervisor abstraction layer are several examples.
```
I'm not sure there's a canon of these deviations

Author: Done.

Each such deviation must be explicitly motivated in the relevant API
documentation or [VEP](https://github.com/kubevirt/enhancements); the mere fact
that something is convenient or already implemented is not sufficient
justification. For example, the [sticky-spec
late-defaulting](#22-identity-persistence-considerations) is a deliberate
deviation from the Kubernetes spec/status split — the VM controller writes
generated hardware identity values back into `VM.spec`, the same resource the

> user manages, rather than into `status`.

Reviewer 2: can you add the justification for this deviation? the important thing about this example is to show future readers how a reasonable justification looks like.

Reviewer 2: or point to https://github.com/kubevirt/kubevirt/pull/18612/changes#diff-ed1897aa85abceb88d0ea13ebbf17c0edbe129222cff2721bbadec319c05aedfR139 ?

Author: Added a pointer to §2.2.1, which already has the justification — didn't want to duplicate it here.

- **Pattern 2 — VM Controller Back-fill (migration pattern):** When a new field
  requiring lifecycle-persistent identity is introduced, existing VMs in the
  cluster will not have it set. If KubeVirt can reproduce the value
  deterministically, the VM controller can compute it and patch it into
  `VM.spec` retroactively. The firmware UUID is again the canonical example: for
  pre-existing VMs that predate the admission webhook logic, the VM controller
  computes the UUID deterministically from the VM name and writes it into
  `VM.spec`.

- **Pattern 3 — Write-Back from Runtime (future consideration):** Some identity
  values are generated by the hypervisor and are challenging to reproduce
  deterministically within KubeVirt (e.g., vNIC MAC addresses or PCI device
  addressing, which must remain stable across reboots to avoid network
  configuration changes). This pattern is not yet implemented generically in
  KubeVirt.

When a new spec field is introduced, its design needs to take into account its
persistence requirements. If it is identified as requiring one, the required
lifecycle (VM vs VMI) must be considered upfront and an explicit persistence
strategy defined.

---

> ### 2.3 Ownership, GC, and Finalizers

Reviewer 2: ```suggestion
### 2.3 Ownership, Garbage Collection, and Finalizers
```

  cloud-native strings (e.g., expressing memory as `2Gi` or disk size as `40Gi`
  instead of raw integers).

- **Fractional or Decimal Weights:** Raw floating-point numbers are strictly
  forbidden within `spec` and strongly discouraged in `status`. Floats introduce
  non-deterministic serialization behavior across different client architectures
  (Go, Python, JavaScript), causing values to silently change when round-tripped
  through different implementations. Floats in `status` additionally cause
  spurious etcd writes when values oscillate in their least significant bits
  between reconciliation cycles. Any field requiring fractional granularity must
  be represented as an integer using milli-units (e.g., representing a
  performance weight of `1.5` as `1500m`); the conversion to the backend's
  native float happens at the translation layer (§5.1). Consider whether a
  frequently changing float value (e.g., byte counters, utilization ratios) is
  better reported via a metrics endpoint — each `status` update is persisted to
  etcd and is not designed for high-frequency telemetry.

  > **Note:** The prohibition is stronger for `spec` than for `status` because

>   > `spec` is written by multiple actors across different languages and client

Reviewer 2: ```suggestion
  > `spec` is typically written by multiple actors across different languages and client
```

Author: Done.

  > implementations — a float written by one client and read back by another may
  > silently change due to serialization differences, and that corrupted value
  > then drives controller logic. `status` is written only by controllers
  > (typically a single Go binary), so the round-trip corruption risk is lower;
  > the remaining concerns are efficiency (spurious etcd writes) and client-side
  > display inconsistency.

- **Oversized Unsigned Values:** Hypervisor configurations that naturally use
  `uint64` must be mapped to safe `string` types or verified `int64` fields
  accompanied by rigid OpenAPI upper/lower boundary validation tags to prevent
  overflow exploits.

---

### 3.2 Open String Set (Avoiding Enums)

#### 3.2.1 Kubernetes Compatibility

> In accordance with upstream [Kubernetes

Reviewer 2: ```suggestion
In accordance with [Kubernetes
```

Author: Done.

  virtual CPU counts (a runtime environment cannot execute on zero or negative
  processing cores).
- **Network & Protocol Specs:** `// +kubebuilder:validation:Maximum=65535` for
  standard TCP/UDP port allocations.
- **Fixed Cryptographic/IEEE Formats:** `// +kubebuilder:validation:Pattern=...`
  for MAC addresses or standard UUID structures where the string layout is
  governed by immutable external RFC standards.

### 4.2 CEL Validation

#### 4.2.1 Prefer CEL over Webhooks

Cross-field validation — where the validity of one configuration parameter
depends directly on the state of another — must prioritize high-performance
**CEL (Common Expression Language)** rules via the
`+kubebuilder:validation:XValidation`
[marker](https://book.kubebuilder.io/reference/markers/crd-validation), which
instructs Kubebuilder to generate the corresponding CEL [validation
rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules)
in the CRD schema. Writing or extending imperative validating admission webhooks
should be treated as a secondary fallback reserved exclusively for highly

> dynamic, external database lookups.

Reviewer 2: Not sure to what you refer by  "external database lookups". A webhook must not refer to those, as this would introduce a TOCTOU race condition. 
These should be warned against, not prescribed.

Author: I tried to phrase it a bit different.
I agree about the limitations of accessing external data, it breaks in many cases the reconciliation idea of the system reaching stable state and self healing. However, there are cases where this is a reasonable compromise.
I think that with the priority outlined and the warning in place, we can balance it correctly.

Of you prefer to place a stronger statement, let me know.

> **Note:** KubeVirt's existing validation logic lives predominantly in

> > imperative admission webhooks, which predate this requirement. These must not

Reviewer 2: ```suggestion
> imperative admission webhooks, which predate CEL and this requirement. These must not
```
Let's be fair by mentioned that CEL is much newer.

Author: Done.

  `Disk` struct has many more backend types; a production rule must account for
  all of them). In CEL, `has(a) != has(b)` is a boolean XOR — it evaluates to
  `true` when exactly one of the two fields is present, and `false` when both or
  neither are present:
  `// +kubebuilder:validation:XValidation:rule="has(self.containerDisk) != has(self.persistentVolumeClaim)",message="A disk slot must define exactly one backend provider (ContainerDisk or PVC)"`

- **Dependency Enforcement:** Verifying that a feature flag is only set when its
  prerequisite is also configured. For example, `isolateEmulatorThread` requires
  `dedicatedCPUPlacement`:

  ```yaml
  # Valid
  spec:
    domain:
      cpu:
        dedicatedCPUPlacement: true
        isolateEmulatorThread: true

  # Rejected: isolation requires dedicated CPU placement
  spec:
    domain:
      cpu:
        dedicatedCPUPlacement: false
        isolateEmulatorThread: true
  ```

  ```go
  // +kubebuilder:validation:XValidation:rule="!has(self.isolateEmulatorThread) || !self.isolateEmulatorThread || self.dedicatedCPUPlacement",message="isolateEmulatorThread requires dedicatedCPUPlacement"

>   ```

Reviewer 2: maybe add: a VMI that is controlled by a VM cannot be modified.

Author: I have already saw implementation in which the VMI is mutated despite managed by a VM object.
I would stick with simple examples.

Reviewer 2: To me it seems like the simplest of examples...
If we allow VMI modification, why do we have https://github.com/kubevirt/kubevirt/pull/18612/changes#diff-ed1897aa85abceb88d0ea13ebbf17c0edbe129222cff2721bbadec319c05aedfR133-R134 ?

Author: That section talks about users, external to Kubevirt.
It mentions:

> only KubeVirt's internal controllers are permitted to modify it.

IMO the VMI protection is a bit more complex, because it conditions the mutability to the Kubevirt SA.

Reviewer 2: gotcha, it's not about the change, but about the user making it.

in the cluster *before* the corresponding controller binaries are fully rolled
out, developers must observe strict safety rules when introducing rules:

- **Safe CEL Usage (New Fields & Pure Invariants):** It is safe to apply CEL
  constraints to newly introduced API fields or to validate absolute, unchanging
  architectural invariants. Older active clients will not be interacting with
  these new properties and are thus insulated from validation triggers.

- **Unsafe CEL Usage (Retroactive Restrictions):** Developers must not introduce
  a CEL rule that retroactively tightens validation criteria on an *existing,
  older field* if an active legacy client or third-party operator could still be
  writing payloads that would violate the new rule — doing so breaks cluster
  state continuity and triggers abrupt, unexpected rejections during the rolling
  upgrade window. **Exception:** Kubernetes 1.28+ supports a CEL evaluation mode
  where a constraint is only checked when the field's value actually changes;
  existing objects with unchanged values continue to pass regardless of whether
  they satisfy the new rule. When the minimum supported cluster version allows

>   it, this mode is the safe way to tighten validation on an existing field.

Reviewer 2: ```suggestion
  it, this mode is the safer way to tighten validation on an existing field.
```
It would still disturb preexisting users.

Author: Done.

- **Specialized Exception:** Highly focused, narrow `Reason` codes are
  acceptable *only* if a consuming controller needs to execute a distinctly
  different, automated remediation pathway for that specific sub-fault.
- **No Multi-Reason Packing:** A condition must never attempt to combine
  multiple distinct failures into a single programmatic reason string (e.g.,
  avoid `StorageAndNetworkNotReady`).

#### 6.2.2 The `Message` Field

The `Message` field is a dynamic, human-readable string intended for end-user
triage, CLI output (`kubectl describe`), and log analysis. All granular
multi-faceted error strings, raw hardware stack traces, nested error codes, or
contextual virtualization parameters must be offloaded entirely to the `Message`
field.

---

### 6.3 Level-Based Status

"Level-based" means the controller acts on the *current* observed state of the
world (the level), rather than reacting to state-change events (edge-triggered).
This is the design pattern recommended by the [Kubernetes API
Conventions](https://github.com/kubernetes/community/blob/main/contributors/devel/sig-architecture/api-conventions.md).

> #### 6.3.1 Recomputation from Scratch

Reviewer 2: that's very important, but it repeats https://github.com/kubevirt/kubevirt/pull/18612/changes#diff-ed1897aa85abceb88d0ea13ebbf17c0edbe129222cff2721bbadec319c05aedfR555-R561 
Maybe you can mention "as discussed earlier" or rewrite the 6.3 without it.

Author: Fair, trimmed 6.3.1 to state the general rule once and point to §5.2.3 for the hypervisor-specific instance instead of restating it.

  recognize the status as stale, defer execution, and wait for the reconciling
  controller to catch up.
- **Speculative/Low-Stakes Logic:** If the consumer is executing non-destructive
  monitoring, speculative routing, or path optimization, it may choose to act
  immediately on the current data, fully aware of the active reconciliation gap.

---

### 6.6 Status Ownership

#### 6.6.1 Single Controller Owner

To eliminate etcd write conflicts, optimistic concurrency failures, and infinite
controller update cascades, the status should be owned by exactly **one**
controller binary. For new resources, this must be enforced structurally by
declaring a [`/status`
subresource](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#status-subresource)
in the CRD. With a status subresource in place, the API server silently discards
status fields included in a spec (main-resource) write — status can only be
updated via the `/status` subresource endpoint — preventing spec writes from
accidentally clobbering observed state. Designing single-ownership in from the
start is always preferable to relying on convention.

> #### 6.6.2 VMI: Dual-Write Debt

Reviewer 2: ```suggestion
#### 6.6.2 VMI: Dual-Writer Debt
```

3. **Define absence semantics explicitly.** `Unknown` — and an absent
    condition, which consumers should treat as `Unknown` — legitimately covers
    two interpretations: the state has not yet been determined, or the condition
    is not relevant to this resource or its current configuration. When
    introducing a new condition, document which of these applies.

#### 6.7.2 Naming Conventions

New condition types should follow the established patterns:

- **Adjective or past-participle for persistent state:** `Ready`, `Paused`,
  `LiveMigratable`, `StorageLiveMigratable`
- **Noun phrase for a pending requirement:** `RestartRequired`,
  `MigrationRequired`, `ManualRecoveryRequired`
- **Noun phrase for an in-progress change:** `HotVCPUChange`, `HotMemoryChange`,
  `VolumesChange`

Avoid negative forms (e.g., `AgentVersionNotSupported` is a legacy outlier —
prefer the positive framing with `status: "False"` and an appropriate `Reason`
to signal the negative case).

#### 6.7.3 Existing Condition Types

The source of truth is
[`staging/src/kubevirt.io/api/core/v1/types.go`](https://github.com/kubevirt/kubevirt/blob/v1.8.4/staging/src/kubevirt.io/api/core/v1/types.go).

> The table below is a representative reference, not exhaustive.

Reviewer 2: why are you repeating it here, then? Do you have good/bad examples in these lists?

Author: I don't think there's a strong reason to keep a duplicate list here that will only drift from types.go.

Dropped both tables, kept the pointer to source.

**VirtualMachineInstance conditions:**

| Type | Semantics |
|----|----|
| `Ready` | VMI is ready to service requests |
| `Paused` | VMI has been paused by the user |
| `Provisioning` | VMI is waiting for one or more DataVolumes to move out of a pending or WaitForFirstConsumer state |
| `Synchronized` | VMI spec is synchronized with the running domain |
| `LiveMigratable` | VMI is eligible for live migration |
| `StorageLiveMigratable` | VMI is eligible for storage live migration |
| `AgentConnected` | Guest agent is connected |
| `AccessCredentialsSynchronized` | Access credentials have been applied successfully |
| `DataVolumesReady` | All DataVolumes attached to the VMI are ready |
| `HotVCPUChange` | A vCPU hot-plug or hot-unplug operation is in progress |
| `HotMemoryChange` | A memory hot-plug or hot-unplug operation is in progress |
| `VolumesChange` | A volume set update is in progress |
| `MigrationRequired` | An automatic migration has been triggered |
| `EvictionRequested` | An eviction has been requested for this VMI |
| `SoftwareEmulation` | The VM is running under software emulation rather than hardware virtualization |

> 

Reviewer 2: We may want to postpone this for a future PR, but I think that we should discuss annotations, their purpose,  benefits and harms verses fully-qualified CRD fields. When should they be used, how they should be named alpha/beta and deprecation...

Author: I agree, it is a worthy topic.
I added it to #18811 .

---

## 7. References & Sources

This document synthesizes rules and patterns from the sources below. Inline
citations throughout link to specific sections where applicable; this list
provides the complete set of sources consulted.

> **Codebase alignment note:** The implementation patterns, examples, and known
> architectural constraints in this document are also grounded in the [KubeVirt
> codebase](https://github.com/kubevirt/kubevirt) as it existed at the time of
> writing. The codebase evolves independently of this document. When assessing
> whether this document remains accurate — whether by a human reviewer or an AI
> agent — the current state of the codebase should be consulted alongside the
> upstream sources listed below to verify that described patterns have not been
> superseded.

> ### Upstream Kubernetes

Reviewer 2: ```suggestion
### Kubernetes
```
Is there a non-upstream Kubernetes?
